manager/manager.go

```diff
@@ -0,0 +1,1095 @@
+		}
+	}
+	return nil, false
+}
+
+// GetProgram - Return a pointer to the requested eBPF program
+// section: section of the program, as defined by its section SEC("[section]")
+// UID: unique identifier given to a probe. If UID is empty, then all the programs matching the provided section are
+// returned.
+func (m *Manager) GetProgram(id ProbeIdentificationPair) ([]*ebpf.Program, bool, error) {
+	var programs []*ebpf.Program
+	if m.collection == nil {
+		return nil, false, ErrManagerNotStarted
+	}
+	if id.UID == "" {
+		for _, probe := range m.Probes {
+			if probe.Section == id.Section {
+				programs = append(programs, probe.program)
+			}
+		}
+		if len(programs) > 0 {
+			return programs, true, nil
+		}
+		prog, ok := m.collection.Programs[id.Section]
+		return []*ebpf.Program{prog}, ok, nil
+	}
+	for _, probe := range m.Probes {
+		if probe.IdentificationPairMatches(id) {
+			return []*ebpf.Program{probe.program}, true, nil
+		}
+	}
+	return programs, false, nil
+}
+
+// GetProgramSpec - Return a pointer to the requested eBPF program spec
+// section: section of the program, as defined by its section SEC("[section]")
+// UID: unique identifier given to a probe. If UID is empty, then the original program spec with the right section in the
+// collection spec (if found) is return
+func (m *Manager) GetProgramSpec(id ProbeIdentificationPair) ([]*ebpf.ProgramSpec, bool, error) {
+	var programs []*ebpf.ProgramSpec
+	if m.collectionSpec == nil {
+		return nil, false, ErrManagerNotInitialized
+	}
+	if id.UID == "" {
+		for _, probe := range m.Probes {
+			if probe.Section == id.Section {
+				programs = append(programs, probe.programSpec)
+			}
+		}
+		if len(programs) > 0 {
+			return programs, true, nil
+		}
+		prog, ok := m.collectionSpec.Programs[id.Section]
+		return []*ebpf.ProgramSpec{prog}, ok, nil
+	}
+	for _, probe := range m.Probes {
+		if probe.IdentificationPairMatches(id) {
+			return []*ebpf.ProgramSpec{probe.programSpec}, true, nil
+		}
+	}
+	return programs, false, nil
+}
+
+// GetProbe - Select a probe by its section and UID
+func (m *Manager) GetProbe(id ProbeIdentificationPair) (*Probe, bool) {
+	for _, managerProbe := range m.Probes {
+		if managerProbe.IdentificationPairMatches(id) {
+			return managerProbe, true
+		}
+	}
+	return nil, false
+}
+
+// CompileAndInit - Start by compiling the provided assets and then initialize the manager.
+func (m *Manager) CompileAndInit(path string, managerOptions Options) error {
+	// TODO: compile the eBPF program and then call Init
+	return nil
+}
+
+// Init - Initialize the manager.
+// elf: reader containing the eBPF bytecode
+func (m *Manager) Init(elf io.ReaderAt) error {
+	return m.InitWithOptions(elf, Options{})
+}
+
+// InitWithOptions - Initialize the manager.
+// elf: reader containing the eBPF bytecode
+// options: options provided to the manager to configure its initialization
+func (m *Manager) InitWithOptions(elf io.ReaderAt, options Options) error {
```

Comment:
I might suggest not using a pointer receiver here and `Init` (perhaps rename to `InitManagerFromReader`), but rather return a `*Manager`. This mirrors how a lot of Go libraries do initialization.

```suggestion
func InitManagerFromReaderWithOptions(elf io.ReaderAt, options Options) (*Manager, error) {
```

Reply:
Good point, however I think it would make sense only if the function was actually creating the manager. If you look at the [examples](https://github.com/DataDog/ebpf/tree/will/manager/manager/examples/programs), the manager is meant to be initialized manually in each package with the entire list of probes, maps and perfmaps (it's essentially following the design of the Cobra library that we use to parse CLI arguments).

The reason for that design is that the Manager options are already quite complex, and we wanted to isolate `runtime options` (like constants editors, map routers, tail call routers, list of activated probes, ...) that might even change over time, from `static options` (essentially the description of what is to be expected in the elf file). Does it make sense ?

Reply:
Yeah, that makes sense. One danger here is misusing the manager accidentally if you don't call `Init`.

Reply:
Absolutely, that's why the `Probes` have a `state` attribute with a lock to make sure that they are properly initialized. Similarly, the `Get*` functions of manager return an error if it was not properly initialized or started. But now that you mention it, I think we should add the same `state` attribute to the manager itself, as well as the maps to ensure that they are not misused either. I'll add something tomorrow.

Reply:
Just added a lock to the Manager and Maps / PerfMaps.